pkg/frontend/vcs/source/find.go

```diff
@@ -22,52 +26,119 @@ type VCSClient interface {
 		client:     client,
 		logger:     logger,
 		repo:       repo,
-		path:       path,
+		file:       file,
 		rootPath:   rootPath,
 		ref:        ref,
 		httpClient: httpClient,
 	}
 }
 
 // Find returns the file content and URL.
-func (ff FileFinder) Find(ctx context.Context) (*vcsv1.GetFileResponse, error) {
-	switch filepath.Ext(ff.path) {
-	case ExtGo:
+func (ff *FileFinder) Find(ctx context.Context) (*vcsv1.GetFileResponse, error) {
+	// first try to gather the config
+	ff.loadConfig(ctx)
+
+	// without config we are done here
+	if ff.config == nil {
+		return ff.findFallback(ctx)
+	}
+
+	// find matching mappings
+	mapping := ff.config.FindMapping(ff.file)
+	if mapping == nil {
+		return ff.findFallback(ctx)
+	}
+
+	switch config.Language(mapping.Language) {
+	case config.LanguageGo:
+		return ff.findGoFile(ctx, mapping)
+	case config.LanguageJava:
+		return ff.findJavaFile(ctx, mapping)
+	// todo: add more languages support
+	default:
+		return ff.findFallback(ctx)
+	}
+}
+
+func (ff FileFinder) findFallback(ctx context.Context) (*vcsv1.GetFileResponse, error) {
+	switch filepath.Ext(ff.file.Path) {
+	case ExtGo, ExtAsm:
 		return ff.findGoFile(ctx)
 	// todo: add more languages support
 	default:
 		// by default we return the file content at the given path without any processing.
-		content, err := ff.fetchRepoFile(ctx, ff.path, ff.ref)
+		content, err := ff.fetchRepoFile(ctx, ff.file.Path, ff.ref)
 		if err != nil {
 			return nil, err
 		}
 		return newFileResponse(content.Content, content.URL)
 	}
 }
 
+// loadConfig attempts to load .pyroscope.yaml from the repository root
+func (ff *FileFinder) loadConfig(ctx context.Context) {
+	sp, ctx := opentracing.StartSpanFromContext(ctx, "FileFinder.loadConfig")
+	defer sp.Finish()
+
+	configPath := config.PyroscopeConfigPath
+	if ff.rootPath != "" {
+		configPath = filepath.Join(ff.rootPath, config.PyroscopeConfigPath)
+	}
+
+	file, err := ff.client.GetFile(ctx, client.FileRequest{
+		Owner: ff.repo.GetOwnerName(),
+		Repo:  ff.repo.GetRepoName(),
+		Path:  configPath,
+		Ref:   ff.ref,
+	})
+	if err != nil {
+		// Config is optional, so just log and continue
+		level.Debug(ff.logger).Log("msg", "no .pyroscope.yaml found", "path", configPath)
+		return
+	}
+	sp.SetTag("config.url", file.URL)
+	sp.SetTag("config", file.Content)
+
+	cfg, err := config.ParsePyroscopeConfig([]byte(file.Content))
+	if err != nil {
+		level.Warn(ff.logger).Log("msg", "failed to parse .pyroscope.yaml", "err", err)
+		return
+	}
+
+	ff.config = cfg
+	level.Debug(ff.logger).Log("msg", "loaded .pyroscope.yaml", "url", file.URL, "mappings", len(cfg.SourceCode.Mappings))
+	sp.SetTag("config.source_code.mappings_count", len(cfg.SourceCode.Mappings))
+
+}
+
 // fetchRepoFile fetches the file content from the configured repository.
 func (arg FileFinder) fetchRepoFile(ctx context.Context, path, ref string) (*vcsv1.GetFileResponse, error) {
+	if arg.rootPath != "" {
+		path = filepath.Join(arg.rootPath, path)
```

Comment:
nit: There's sort of a rough edge here where this root path is defined in the UI at repo level and then the mapping config can contain local paths that are function-level root paths. I can see as a user being confused playing with these two configs.

Maybe at some point a helper tooltip in the UI can clarify how these configs interact.